### PR TITLE
Refactor charting

### PR DESCRIPTION
**Pull request by DevAndyLee**
Refactor of charting to re-use charting code, and implement it as a custom knockout binding.

Changed Ranked chart to a per-round display that will hopefully be easier to read and understand (and doesn't need the filtering options to make sense of it).

Added a "Winner" label to make it really obvious who won

### CODE REVIEW

File: VotingApplication/VotingApplication.Web.Api/Scripts/BasicVote.js [hunk 1]
```diff
@@ -83,61 +86,20 @@
...
-            .tooltipFunction(function (d) {
-                var maxToDisplay = 5;
-                if (d.Count <= maxToDisplay) {
-                    return "Votes: " + d.Count + "<br />" + d.Voters.toString().replace(/,/g, "<br />");
-                }
-                else {
-                    return "Votes: " + d.Count + "<br />" + d.Voters.slice(0, maxToDisplay).toString().replace(/,/g, "<br />") + "<br />" + "+ " + (d.Count - maxToDisplay) + " others";
+            var winner = { Name: "", Sum: 0 };
+            groupedVotes.forEach(function (optionVotes) {
+                if (optionVotes.Sum > winner.Sum) {
```
Comment (tpkelly) [inline VotingApplication/VotingApplication.Web.Api/Scripts/BasicVote.js]: We need to let the voters know if there is a tie too

File: VotingApplication/VotingApplication.Web.Api/Scripts/PointsVote.js [hunk 2]
```diff
@@ -127,7 +130,16 @@
 
         self.displayResults = function(data) {
             var groupedVotes = countVotes(data);
-            self.drawChart(groupedVotes);
+            self.chartData([{ Data: groupedVotes }]);
+
+            var winner = { Name: "", Sum: 0 };
+            groupedVotes.forEach(function (optionVotes) {
+                if (optionVotes.Sum > winner.Sum) {
+                    winner = optionVotes;
```
Comment (tpkelly) [inline VotingApplication/VotingApplication.Web.Api/Scripts/PointsVote.js]: Same again

File: VotingApplication/VotingApplication.Web.Api/Scripts/ResultChart.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        var chart;
+
+        this.drawChart = function (data, options) {
+            if (!data.length) return;
+
+            var dataUnchanged = false;
+            var barCount = 0;
+            for (var n = 0; n < data.length; n++) {
+                barCount += data[n].Data.length;
+                dataUnchanged = dataUnchanged &&
```
Comment (tpkelly) [inline VotingApplication/VotingApplication.Web.Api/Scripts/ResultChart.js]: If we know early on that the data has changed, we should break out of the for loop immediately.

File: VotingApplication/VotingApplication.Web.Api/Scripts/RankedVote.js [hunk 4]
```diff
@@ -242,121 +244,6 @@
...
-                    else
-                        return "Option eliminated";
-                });
-
-                var newSeries = chart.series();
-                newSeries.push(series);
-                chart.series(newSeries);
-            });
-
-            //Annotate the decision line
```
Comment (tpkelly) [inline VotingApplication/VotingApplication.Web.Api/Scripts/RankedVote.js]: The annotation is important for the Ranked poll. We should still be able to add a line at voterCount/2.

File: VotingApplication/VotingApplication.Web.Api/Scripts/ResultChart.js [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                columns: false
+            };
+            var settings = $.extend(defaultSettings, options);
+
+            // Hack to fix insight's lack of data reloading
+            $element.html('');
+
+            chart = new insight.Chart('', $element[0])
+                .width($element.width())
+                .height(400); //Math.min(barCount * 50 + 100, 400));
```
Comment (tpkelly) [inline VotingApplication/VotingApplication.Web.Api/Scripts/ResultChart.js]: The bars look a bit big when there are not many options, and squashed when there are. Can we scale this by the number of options/rounds?

**Comment by DevAndyLee** [comment]
Comments have been addressed by the last two commits

**Pull request merged**